File: code/45.jump-game-ii.java
/*
 * @lc app=leetcode id=45 lang=java
 *
 * [45] Jump Game II
 */

// @lc code=start
class Solution {
    public int jump(int[] nums) {
        if (nums == null || nums.length == 0 || nums.length == 1) {
            return 0;
        }
        int currCover = 0;
        int maxCover = 0;
        int res = 0;
        for (int i = 0; i < nums.length; i++) {
            maxCover = Math.max(i + nums[i], maxCover);
            if (maxCover >= nums.length - 1) {
                res++;
                break;
            }

            if (i == currCover) {
                currCover = maxCover;
                res++;
            }
        }
        return res;
    }
}
// @lc code=end

File: code/45.jump-game-ii.go
/*
 * @lc app=leetcode id=45 lang=golang
 *
 * [45] Jump Game II
 */

// @lc code=start
func jump(nums []int) int {
	if nums == nil || len(nums) == 0 || len(nums) == 1 {
		return 0
	}

	res, maxCover, currCover := 0, 0, 0
	for i := 0; i < len(nums); i++ {
		if (i + nums[i]) > maxCover {
			maxCover = i + nums[i]
		}

		if maxCover >= (len(nums) - 1) {
			res++
			break
		}

		if i == currCover {
			currCover = maxCover
			res++
		}
	}
	return res
}

// @lc code=end

